src/lib/OAuth2/OAuth2.ts

```diff
@@ -0,0 +1,209 @@
+  loadUserInfo?: boolean;
+  revokeAccessTokenOnLogout?: boolean;
+  filterProtocolClaims?: boolean;
+  validateSubOnSilentRenew?: boolean;
+  persistenceMethod?: Storage;
+  customMetadata?: Partial<IOAuth2CustomMetadata>;
+  signingKeys?: IOAuth2SigningKey[];
+}
+
+class OAuth2 {
+  protected userManager: UserManager;
+  protected eventEmitter: EventTarget;
+
+  constructor(config: IOAuth2Config) {
+    const persistenceMethod = new WebStorageStateStore({
+      store: config.persistenceMethod,
+    });
+
+    const customMetadata = convertToOIDCMetadata(config.customMetadata);
+
+    const managerConfig: UserManagerSettings = {
+      authority: config.authority,
+      client_id: config.clientId,
+      client_secret: config.clientSecret,
+      redirect_uri: config.redirectUri,
+      response_type: config.responseType,
+      response_mode: config.responseMode,
+      scope: config.scope,
+      prompt: config.prompt,
+      automaticSilentRenew: config.automaticSilentRenew,
+      includeIdTokenInSilentRenew: config.includeIDTokenInSilentRenew,
+      loadUserInfo: config.loadUserInfo,
+      revokeAccessTokenOnSignout: config.revokeAccessTokenOnLogout,
+      filterProtocolClaims: config.filterProtocolClaims,
+      validateSubOnSilentRenew: config.validateSubOnSilentRenew,
+      userStore: persistenceMethod,
+      signingKeys: config.signingKeys,
+      metadata: customMetadata,
+    };
+
+    this.eventEmitter = new EventTarget();
+    this.userManager = new UserManager(managerConfig);
+
+    this.registerInternalEvents();
+  }
+
+  public attemptLogin(): Promise<void> {
+    return this.userManager.signinRedirect({ useReplaceToNavigate: true });
+  }
+
+  public async handleLoginResponse(
+    currentURL: string
+  ): Promise<OAuth2UserImpl> {
+    const origUser = await this.userManager.signinRedirectCallback(currentURL);
+    const newUser = OAuth2UserImpl.fromOIDCUser(origUser);
+
+    return newUser;
+  }
+
+  public async getLoggedInUser(): Promise<OAuth2UserImpl | null> {
+    let origUser = await this.userManager.getUser();
+    if (!origUser) return null;
+
+    // If user is already expired, renew his authentication.
+    if (origUser.expired) {
+      origUser = await this.userManager.signinSilent();
+    }
+
+    this.userManager.events.load(origUser);
+    const newUser = OAuth2UserImpl.fromOIDCUser(origUser);
+
+    return newUser;
+  }
+
+  public async renewUser(): Promise<OAuth2UserImpl> {
+    const origUser = await this.userManager.signinSilent();
+
+    this.userManager.events.load(origUser);
+    const newUser = OAuth2UserImpl.fromOIDCUser(origUser);
+
+    return newUser;
+  }
+
+  public async logout(): Promise<void> {
+    await this.userManager.removeUser();
+    await this.userManager.clearStaleState();
+  }
+
+  public addEventListener<
+    T extends OAuth2Events,
+    U extends IOAuth2EventCallbacks[T]
+  >(event: T, cb: U) {
+    this.eventEmitter.addEventListener(event, cb as EventListener, false);
```

Comment:
I forgot to mention that using this approach will auto-suggest the right callback type for each type of event, based on the `IOAuth2EventCallbacks` interface.